caddytls/storagetest/storagetest.go

```diff
@@ -0,0 +1,304 @@
+		if _, err := loadFn(key); err != caddytls.ErrStorageNotFound {
+			return fmt.Errorf("Expected ErrStorageNotFound after delete, got: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// TestSiteCert tests Storage.*SiteCert.
+func (s *StorageTest) TestSiteCert() error {
+	if err := s.runPreTest(); err != nil {
+		return err
+	}
+	defer s.runPostTest()
+	return testLoadAndStore("example.com", s.LoadSiteCert, s.StoreSiteCert, s.DeleteSiteCert)
+}
+
+// TestSiteKey tests Storage.*SiteKey.
+func (s *StorageTest) TestSiteKey() error {
+	if err := s.runPreTest(); err != nil {
+		return err
+	}
+	defer s.runPostTest()
+	return testLoadAndStore("example.com", s.LoadSiteKey, s.StoreSiteKey, nil)
+}
+
+// TestSiteMeta tests Storage.*SiteMeta.
+func (s *StorageTest) TestSiteMeta() error {
+	if err := s.runPreTest(); err != nil {
+		return err
+	}
+	defer s.runPostTest()
+	return testLoadAndStore("example.com", s.LoadSiteMeta, s.StoreSiteMeta, nil)
+}
+
+// TestUserReg tests Storage.*UserReg.
+func (s *StorageTest) TestUserReg() error {
+	if err := s.runPreTest(); err != nil {
+		return err
+	}
+	defer s.runPostTest()
+	return testLoadAndStore("foo@example.com", s.LoadUserReg, s.StoreUserReg, nil)
+}
+
+// TestUserKey tests Storage.*UserKey.
+func (s *StorageTest) TestUserKey() error {
+	if err := s.runPreTest(); err != nil {
+		return err
+	}
+	defer s.runPostTest()
+	return testLoadAndStore("foo@example.com", s.LoadUserKey, s.StoreUserKey, nil)
+}
+
+// TestMostRecentUserEmail tests Storage.MostRecentUserEmail.
+func (s *StorageTest) TestMostRecentUserEmail() error {
+	if err := s.runPreTest(); err != nil {
+		return err
+	}
+	defer s.runPostTest()
+
+	// Should be empty on first run
+	if e := s.MostRecentUserEmail(); e != "" {
+		return fmt.Errorf("Expected empty most recent user on first run, got: %v", e)
+	}
+
+	// If we store user reg, then that one should be returned
+	if err := s.StoreUserReg("foo1@example.com", []byte("foo")); err != nil {
+		return err
+	}
+	if s.AfterUserEmailStore != nil {
+		s.AfterUserEmailStore("foo1@example.com")
+	}
+	if e := s.MostRecentUserEmail(); e != "foo1@example.com" {
+		return fmt.Errorf("Unexpected most recent email after user reg: %v", e)
+	}
+
+	// If we store user key, then that one should be returned
+	if err := s.StoreUserKey("foo2@example.com", []byte("foo")); err != nil {
+		return err
+	}
+	if s.AfterUserEmailStore != nil {
+		s.AfterUserEmailStore("foo2@example.com")
+	}
+	if e := s.MostRecentUserEmail(); e != "foo2@example.com" {
+		return fmt.Errorf("Unexpected most recent email after user key: %v", e)
+	}
+
+	// If we store user reg again, it should NOT should be returned
+	if err := s.StoreUserReg("foo1@example.com", []byte("foo")); err != nil {
+		return err
+	}
+	if s.AfterUserEmailStore != nil {
+		s.AfterUserEmailStore("foo1@example.com")
+	}
+	if e := s.MostRecentUserEmail(); e != "foo1@example.com" {
+		return fmt.Errorf("Unexpected most recent email after second user reg: %v", e)
+	}
+	return nil
+}
```

Comment:
This is a little unconventional; let's move the test logic into a true `_test.go` file.